### PR TITLE
[SPARK-44301][SQL] Add Benchmark Suite for TPCH

### PR DESCRIPTION
**Pull request by oss-maker**
### What changes were proposed in this pull request?
This PR adds support to run benchmark for TPCH. The schema used is defined here in the official TPCH site - https://www.tpc.org/TPC_Documents_Current_Versions/pdf/TPC-H_v3.0.1.pdf

### Why are the changes needed?
This PR makes it easier to run the benchmark for TPCH especially in scenarios where we want to test aggregate performance of TPCH aggregate heavy queries like Query 1

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
This PR was tested locally and verified that the benchmark runs properly.

### CODE REVIEW

**Comment by LuciferYang** [comment]
All benchmarks can be run using Github Action:

https://github.com/apache/spark/blob/master/.github/workflows/benchmark.yml

At the same time, all benchmark results generated by GitHub Action and records in Spark, such as 

https://github.com/apache/spark/blob/master/sql/core/benchmarks/TPCDSQueryBenchmark-results.txt

So I think if we need to introduce `TPCHQueryBenchmark`, it should also be able to run using GitHub Action like others.

**Comment by oss-maker** [comment]
Thanks for the review @LuciferYang, I'll make the change to add this to the Github action.

**Comment by oss-maker** [comment]
@LuciferYang I have added the TPCH Benchmark to Github action, but I don't see any benchmark workflows being triggered (both TPCDS and TPCH), any idea why and how I can trigger them? Thanks

**Comment by oss-maker** [comment]
Triggered a workflow - https://github.com/oss-maker/spark/actions/runs/5466653951

**Comment by srinivasst** [comment]
Lgtm

**Comment by oss-maker** [comment]
@LuciferYang I have done the required change, thanks for your help, could you please review once. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala [hunk 1]
```diff
@@ -20,14 +20,11 @@ package org.apache.spark.sql.execution.benchmark
...
 
 
-class TPCDSQueryBenchmarkArguments(val args: Array[String]) {
-  var dataLocation: String = sys.env.getOrElse("SPARK_TPCDS_DATA", null)
+class TPCQueryBenchmarkArguments(val args: Array[String], val dataLocationEnv: String) {
+  var dataLocation: String = sys.env.getOrElse(dataLocationEnv, null)
   var queryFilter: Set[String] = Set.empty
   var cboEnabled: Boolean = false
 
-  parseArgs(args.toList)
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala:28]: why remove the invoke of `parseArgs` and `validateArguments`, these 2 functions will be unused after this pr

**Review by LuciferYang — COMMENTED** [review]
Need more people participate review, cc @HyukjinKwon @dongjoon-hyun @MaxGekk @wangyum FYI

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCHData.scala [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+
+import scala.util.Try
+
+import org.apache.spark.SparkContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions.{col, rpad}
+import org.apache.spark.sql.types.{CharType, StringType, StructField, StructType, VarcharType}
+
+// The classes in this file are basically moved from https://github.com/databricks/spark-sql-perf
```
[the remark this answers is not in the document]
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCHData.scala:28]: Yes, I have based this on that, note that there is no GenTPCHData, so, that is new

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBenchmarkUtils.scala [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.spark.sql.catalyst.catalog.HiveTableRelation
+import org.apache.spark.sql.catalyst.plans.logical.SubqueryAlias
+import org.apache.spark.sql.catalyst.util.DateTimeConstants.NANOS_PER_SECOND
+import org.apache.spark.sql.catalyst.util.resourceToString
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+
+
+trait TPCBenchmarkUtils extends SqlBasedBenchmark with Logging {
```
[the remark this answers is not in the document]
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBenchmarkUtils.scala]: Thanks, fixed the name

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCHData.scala [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+
+class Dbgen(dbgenDir: String, params: Seq[String]) extends Serializable {
+  val dbgen = s"$dbgenDir/dbgen"
+  def generate(
+    sparkContext: SparkContext,
```
[the remark this answers is not in the document]
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCHData.scala]: Fixed it. Thanks

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala (hunk 1 above)
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala]: My bad, it's fixed now

**Comment by surnaik** [comment]
@beliefer Could you also please review once. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+    )
+    val paramsString = params.mkString(" ")
+    val command = s"./dbgen -q $paramsString -T ${shortTableNames(tableName)} -s $scaleFactor"
+    generateData(sparkContext, tableName, partitions, scaleFactor, command)
+  }
+}
+
+trait TableGenerator extends Serializable with Logging {
+  protected val dataGenerator: DataGenerator
+  protected val sparkSQLContext: SQLContext
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala:164]:
```suggestion
  protected val sqlContext: SQLContext
```

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+  protected val tpcScaleFactor: Int
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala:165]:
```suggestion
  protected val scaleFactor: Int
```

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala [hunk 7]
```diff
@@ -17,391 +17,20 @@
...
 class TPCDSTables(sqlContext: SQLContext, dsdgenDir: String, scaleFactor: Int)
-  extends TPCDSSchema with Logging with Serializable {
+  extends TableGenerator with TPCDSSchema with Logging with Serializable {
 
-  private val dataGenerator = new Dsdgen(dsdgenDir)
-
-  private def tables: Seq[Table] = tableColumns.map { case (tableName, schemaString) =>
+  override protected val dataGenerator: DataGenerator = new Dsdgen(dsdgenDir)
+  override protected val sparkSQLContext: SQLContext = sqlContext
+  override protected val tpcScaleFactor: Int = scaleFactor
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala:28]:
```suggestion
```

Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala:28]: Please declare `sqlContext: SQLContext` and `scaleFactor: Int` in `TableGenerator`.

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+
+  def generate(
+      sparkContext: SparkContext,
+      tableName: String,
+      partitions: Int,
+      scaleFactor: Int): RDD[String]
+}
+
+class Dsdgen(dsdgenDir: String) extends DataGenerator with Serializable {
+  override protected val toolsDir: String = s"$dsdgenDir"
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala:126]: Shall we aovid the duplicated variable `dsdgenDir` and `toolsDir` ?

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala:164]: BTW. SQLContext is an old API. It seems it's the time use `SparkSession` instead. cc @cloud-fan @MaxGekk @wangyum

**Comment by beliefer** [comment]
@surnaik Thank you for you ping.

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala (hunk 7 above)
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala]: Updated. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala (hunk 6 above)
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala]: Updated.

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala (hunk 5 above)
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala]: The sqlContext is extracted out of the same SparkSession, I don't think this is required, even if we pass it, in the genData, we will have to again get hold of the sqlContext or catalog to operate on table. Let me know if this is absolutely required. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala [hunk 9]
```diff
@@ -63,15 +63,15 @@ class TPCDSQueryBenchmarkArguments(val args: Array[String]) {
...
+      |  --data-location      Path to TPCDS/H data
       |  --query-filter       Queries to filter, e.g., q3,q5,q13
       |  --cbo                Whether to enable cost-based optimization
       |
       |------------------------------------------------------------------------------------------------------------------
       |In order to run this benchmark, please follow the instructions at
       |https://github.com/databricks/spark-sql-perf/blob/master/README.md
-      |to generate the TPCDS data locally (preferably with a scale factor of 5 for benchmarking).
-      |Thereafter, the value of <TPCDS data location> needs to be set to the location where the generated data is stored.
+      |to generate the TPCDS/H data locally (preferably with a scale factor of 5 for benchmarking).
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala:73]: It seems a little odd. May it be `TPC-DS/TPC-H`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala [hunk 10]
```diff
@@ -63,15 +63,15 @@ class TPCDSQueryBenchmarkArguments(val args: Array[String]) {
...
+      |Thereafter, the value of <TPCDS/H data location> needs to be set to the location where the generated data is stored.
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala:74]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala [hunk 11]
```diff
@@ -63,15 +63,15 @@ class TPCDSQueryBenchmarkArguments(val args: Array[String]) {
     System.err.println("""
       |Usage: spark-submit --class <this class> <spark sql test jar> [Options]
       |Options:
-      |  --data-location      Path to TPCDS data
+      |  --data-location      Path to TPCDS/H data
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCQueryBenchmarkArguments.scala:66]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+      .set("spark.executor.memory", "3g")
+      .set("spark.sql.autoBroadcastJoinThreshold", (20 * 1024 * 1024).toString)
+      .set("spark.sql.crossJoin.enabled", "true")
+      .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+      .set("spark.kryo.registrationRequired", "true")
+
+    SparkSession.builder.config(conf).getOrCreate()
+  }
+
+  def setupTables(dataLocation: String, tableColumns: Map[String, StructType]): Map[String, Long] =
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala:56]:
```suggestion
  protected def setupTables(dataLocation: String, tableColumns: Map[String, StructType]): Map[String, Long] =
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+      // Recover partitions but don't fail if a table is not partitioned.
+      Try {
+        spark.sql(s"ALTER TABLE $tableName RECOVER PARTITIONS")
+      }.getOrElse {
+        logInfo(s"Recovering partitions of table $tableName failed")
+      }
+      tableName -> spark.table(tableName).count()
+    }.toMap
+
+  def runTpcQueries(
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala:70]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+      val numRows = queryRelations.map(tableSizes.getOrElse(_, 0L)).sum
+      val benchmark = new Benchmark(s"$queryType Snappy", numRows, 2, output = output)
+      benchmark.addCase(s"$name$nameSuffix") { _ =>
+        spark.sql(queryString).noop()
+      }
+      benchmark.run()
+    }
+  }
+
+  def filterQueries(
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala:101]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        origQueries.filter { name => queryFilter.contains(s"$name$nameSuffix") }
+      } else {
+        origQueries.filter(queryFilter.contains)
+      }
+    } else {
+      origQueries
+    }
+  }
+
+  def configureCbo(benchmarkArgs: TPCQueryBenchmarkArguments): Unit = {
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/TPCBasedBenchmark.scala:116]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala]: I created https://github.com/apache/spark/pull/41900 to replace `SQLContext` with `SparkSession` for `GenTPCDSData` and it works good.

@surnaik Could you wait for https://github.com/apache/spark/pull/41900 merged and update this PR ?

Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala]: Sounds good. Thanks, please update once #41900 is merged.

**Comment by oss-maker** [comment]
@beliefer Thanks for the change, I have updated my PR after the merge of #41900

**Comment by oss-maker** [comment]
@beliefer Thanks for your suggestion, I wasn't even aware of this Scala feature. That's why I kept using a different name in the class parameter. I appreciate your help, this was very useful.

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+    val paramsString = params.mkString(" ")
+    val command = s"./dbgen -q $paramsString -T ${shortTableNames(tableName)} -s $scaleFactor"
+    generateData(sparkContext, dbgenDir, tableName, partitions, scaleFactor, command)
+  }
+}
+
+trait TableGenerator extends Serializable with Logging {
+  protected val dataGenerator: DataGenerator
+  protected val spark: SparkSession
+  protected val scaleFactor: Int
```
[the remark this answers is not in the document]
Comment (oss-maker) [reply sql/core/src/test/scala/org/apache/spark/sql/GenTPCData.scala:162]: This is not used in the trait at all, so we don't need the dsdgenDir and dbgenDir to be declared in the trait, they are used to construct the DataGenerator, so we can directly use them in the TPCDSTables and TPCHTables.

File: sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala [hunk 17]
```diff
@@ -17,390 +17,18 @@
...
-    val done: Int => Unit,
-    val stream: () => Stream[T])
+import org.apache.spark.sql.types.StructType
 
-  // See scala.sys.process.Streamed
-  private object BlockingStreamed {
-    // scala.process.sys.Streamed uses default of Integer.MAX_VALUE,
-    // which causes OOMs if the consumer cannot keep up with producer.
-    val maxQueueSize = 65536
+class TPCDSTables(val spark: SparkSession, val dsdgenDir: String, val scaleFactor: Int)
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/GenTPCDSData.scala:23]: `dsdgenDir` is not extended from the parent trait, so remove the `val`.

**Review by beliefer — COMMENTED** [review]
cc @maropu @LuciferYang

**Comment by LuciferYang** [comment]
Thanks @beliefer,, please wait me check this pr with Scala 2.13

**Comment by surnaik** [comment]
@LuciferYang The size of the data is incredibly small at SF=1. The cache is working fine.

**Comment by LuciferYang** [comment]
The data only 4KB when sf = 1?

**Comment by surnaik** [comment]
Ya, that's strange, even I'm very confused about that, using the tpch-dbgen
tool from the github.com/databricks/tpch-dbgen repository, the cache shows
~4KB.

Thanks,
Suraj Naik

**Comment by LuciferYang** [comment]
Should a larger TPCH dataset be used?

Correspondingly, when sf=1, the size of the tpcds dataset  cache is 252MB

**Comment by surnaik** [comment]
Ya, let me try with 5 and check once. Will update here once I have the result.

**Comment by LuciferYang** [comment]
Thanks @surnaik

**Comment by surnaik** [comment]
I tried with SF 10, the actual data size is definitely large but the cache size actually shrunk to ~3.5KB. Let me locally try it once.

File: .github/workflows/benchmark.yml [hunk 18]
```diff
@@ -186,7 +251,7 @@ jobs:
         # To keep the directory structure and file permissions, tar them
         # See also https://github.com/actions/upload-artifact#maintaining-file-permissions-and-case-sensitive-files
         echo "Preparing the benchmark results:"
-        tar -cvf benchmark-results-${{ github.event.inputs.jdk }}-${{ github.event.inputs.scala }}.tar `git diff --name-only` `git ls-files --others --exclude=tpcds-sf-1 --exclude-standard`
+        tar -cvf benchmark-results-${{ github.event.inputs.jdk }}-${{ github.event.inputs.scala }}.tar `git diff --name-only` `git ls-files --others --exclude=tpcds-sf-1 --exclude=tpch-sf-1 --exclude-standard`
```
Comment (LuciferYang) [inline .github/workflows/benchmark.yml:254]: @srinivasst  I remove `--exclude=tpch-sf-1` and download the generated tpch data, then I found that the data file only contains parquet footer and no data

Comment (surnaik) [reply .github/workflows/benchmark.yml:254]: Interesting find, let me check the local data, but do you know where else would tpch dbgen be available from?
I'm using from github.com/databricks/tpch-dbgen

Comment (LuciferYang) [reply .github/workflows/benchmark.yml:254]: Sorry, I don’t know ...

**Comment by surnaik** [comment]
Back from a break. I will use the official dbgen from TPCH website and update the PR. Thanks!

**Comment by pan3793** [comment]
@surnaik Apache Kyuubi has a [Spark TPC-H Connector](https://github.com/apache/kyuubi/tree/master/extensions/spark/kyuubi-spark-connector-tpch/src/main/scala/org/apache/kyuubi/spark/connector/tpch) based on DataSource V2 API, which is inspired by [TPCH connector](https://trino.io/docs/current/connector/tpch.html#connector-tpch--page-root)

```
spark-sql \
  --packages org.apache.kyuubi:kyuubi-spark-connector-tpch_2.12:1.7.1 \
  --conf spark.sql.catalog.tpch=org.apache.kyuubi.spark.connector.tpch.TPCHCatalog
```

```
spark-sql> show tables in tpch.sf1;
customer
orders
lineitem
part
partsupp
supplier
nation
region
Time taken: 0.1 seconds, Fetched 8 row(s)

spark-sql> select * from tpch.sf1.orders limit 10;
1	36901	O	173665.47	1996-01-02	5-LOW	Clerk#000000951	0	nstructions sleep furiously among
2	78002	O	46929.18	1996-12-01	1-URGENT	Clerk#000000880	0	 foxes. pending accounts at the pending, silent asymptot
3	123314	F	193846.25	1993-10-14	5-LOW	Clerk#000000955	0	sly final accounts boost. carefully regular ideas cajole carefully. depos
4	136777	O	32151.78	1995-10-11	5-LOW	Clerk#000000124	0	sits. slyly regular warthogs cajole. regular, regular theodolites acro
5	44485	F	144659.2	1994-07-30	5-LOW	Clerk#000000925	0	quickly. bold deposits sleep slyly. packages use slyly
6	55624	F	58749.59	1992-02-21	4-NOT SPECIFIED	Clerk#000000058	0	ggle. special, final requests are against the furiously specia
7	39136	O	252004.18	1996-01-10	2-HIGH	Clerk#000000470	0	ly special requests
32	130057	O	208660.75	1995-07-16	2-HIGH	Clerk#000000616	0	ise blithely bold, regular requests. quickly unusual dep
33	66958	F	163243.98	1993-10-27	3-MEDIUM	Clerk#000000409	0	uriously. furiously final request
34	61001	O	58949.67	1998-07-21	3-MEDIUM	Clerk#000000223	0	ly final packages. fluffily final deposits wake blithely ideas. spe
Time taken: 2.608 seconds, Fetched 10 row(s)
```

It could also be used to generate the data by using CTAS.

**Pull request closed without merging**